### PR TITLE
[Arm64] ASIMD DotProduct and Rounding Double Multiply Add/Subtract

### LINKED ISSUE #36696: ARM Doubling Multiply intrinsics
```csharp
namespace System.Runtime.Intrinsics.Arm
{
    public abstract class AdvSimd
    {
        /// <summary>
        /// Signed saturating doubling multiply returning high half
        /// Corresponds to vector forms of SQDMULH and VQDMULH
        /// </summary>
        public static Vector64<short> DoublingMultiplySaturateHigh(Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplySaturateHigh(Vector64<int> left, Vector64<int> right);
        public static Vector128<short> DoublingMultiplySaturateHigh(Vector128<short> left, Vector128<short> right);
        public static Vector128<int> DoublingMultiplySaturateHigh(Vector128<int> left, Vector128<int> right);

        public static Vector64<short> DoublingMultiplyBySelectedScalarSaturateHigh(Vector64<short> left, Vector64<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplyBySelectedScalarSaturateHigh(Vector64<int> left, Vector64<int> right, byte rightIndex);
        public static Vector64<short> DoublingMultiplyBySelectedScalarSaturateHigh(Vector64<short> left, Vector128<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplyBySelectedScalarSaturateHigh(Vector64<int> left, Vector128<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplyBySelectedScalarSaturateHigh(Vector128<short> left, Vector64<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplyBySelectedScalarSaturateHigh(Vector128<int> left, Vector64<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplyBySelectedScalarSaturateHigh(Vector128<short> left, Vector128<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplyBySelectedScalarSaturateHigh(Vector128<int> left, Vector128<int> right, byte rightIndex);

        public static Vector64<short> DoublingMultiplySaturateHighScalar(Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplySaturateHighScalar(Vector64<int> left, Vector64<int> right);

        // TODO: Scalar BySelectedScalar variant

        /// <summary>
        /// Signed saturating rounding doubling multiply returning high half
        /// Corresponds to vector forms of SQRDMULH and VQRDMULH
        /// </summary>
        public static Vector64<short> RoundedDoublingMultiplySaturateHigh(Vector64<short> left, Vector64<short> right);
        public static Vector64<int> RoundedDoublingMultiplySaturateHigh(Vector64<int> left, Vector64<int> right);
        public static Vector128<short> RoundedDoublingMultiplySaturateHigh(Vector128<short> left, Vector128<short> right);
        public static Vector128<int> RoundedDoublingMultiplySaturateHigh(Vector128<int> left, Vector128<int> right);

        public static Vector64<short> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector64<short> left, Vector64<short> right, byte rightIndex);
        public static Vector64<int> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector64<int> left, Vector64<int> right, byte rightIndex);
        public static Vector64<short> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector64<short> left, Vector128<short> right, byte rightIndex);
        public static Vector64<int> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector64<int> left, Vector128<int> right, byte rightIndex);
        public static Vector128<short> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector128<short> left, Vector64<short> right, byte rightIndex);
        public static Vector128<int> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector128<int> left, Vector64<int> right, byte rightIndex);
        public static Vector128<short> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector128<short> left, Vector128<short> right, byte rightIndex);
        public static Vector128<int> RoundedDoublingMultiplyBySelectedScalarSaturateHigh(Vector128<int> left, Vector128<int> right, byte rightIndex);

        public static Vector64<short> RoundedDoublingMultiplySaturateHighScalar(Vector64<short> left, Vector64<short> right);
        public static Vector64<int> RoundedDoublingMultiplySaturateHighScalar(Vector64<int> left, Vector64<int> right);

        // TODO: Scalar BySelectedScalar variant

        /// <summary>
        /// Signed Saturating Rounding Doubling Multiply Accumulate returning High Half
        /// Corresponds to vector forms of SQRDMLAH and VQRDMLAH
        /// </summary>
        public static Vector64<short> RoundedDoublingMultiplyAddSaturateHigh(Vector64<short> addend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> RoundedDoublingMultiplyAddSaturateHigh(Vector64<int> addend, Vector64<int> left, Vector64<int> right);
        public static Vector128<short> RoundedDoublingMultiplyAddSaturateHigh(Vector128<short> addend, Vector128<short> left, Vector128<short> right);
        public static Vector128<int> RoundedDoublingMultiplyAddSaturateHigh(Vector128<int> addend, Vector128<int> left, Vector128<int> right);

        public static Vector64<short> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector64<short> addend, Vector64<short> left, Vector64<short> right, byte rightIndex);
        public static Vector64<int> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector64<int> addend, Vector64<int> left, Vector64<int> right, byte rightIndex);
        public static Vector64<short> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector64<short> addend, Vector64<short> left, Vector128<short> right, byte rightIndex);
        public static Vector64<int> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector64<int> addend, Vector64<int> left, Vector128<int> right, byte rightIndex);
        public static Vector128<short> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector128<short> addend, Vector128<short> left, Vector64<short> right, byte rightIndex);
        public static Vector128<int> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector128<int> addend, Vector128<int> left, Vector64<int> right, byte rightIndex);
        public static Vector128<short> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector128<short> addend, Vector128<short> left, Vector128<short> right, byte rightIndex);
        public static Vector128<int> RoundedDoublingMultiplyAddBySelectedScalarSaturateHigh(Vector128<int> addend, Vector128<int> left, Vector128<int> right, byte rightIndex);

        public static Vector64<short> RoundedDoublingMultiplyAddSaturateHighScalar(Vector64<short> addend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> RoundedDoublingMultiplyAddSaturateHighScalar(Vector64<int> addend, Vector64<int> left, Vector64<int> right);

        // TODO: Scalar BySelectedScalar variant

        /// <summary>
        /// Signed saturating rounding doubling multiply subtract returning high half
        /// Corresponds to vector forms of SQRDMLSH and VQRDMLSH
        /// </summary>
        public static Vector64<short> RoundedDoublingMultiplySubtractSaturateHigh(Vector64<short> minuend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> RoundedDoublingMultiplySubtractSaturateHigh(Vector64<int> minuend, Vector64<int> left, Vector64<int> right);
        public static Vector128<short> RoundedDoublingMultiplySubtractSaturateHigh(Vector128<short> minuend, Vector128<short> left, Vector128<short> right);
        public static Vector128<int> RoundedDoublingMultiplySubtractSaturateHigh(Vector128<int> minuend, Vector128<int> left, Vector128<int> right);

        public static Vector64<short> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector64<short> minuend, Vector64<short> left, Vector64<short> right, byte rightIndex);
        public static Vector64<int> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector64<int> minuend, Vector64<int> left, Vector64<int> right, byte rightIndex);
        public static Vector64<short> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector64<short> minuend, Vector64<short> left, Vector128<short> right, byte rightIndex);
        public static Vector64<int> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector64<int> minuend, Vector64<int> left, Vector128<int> right, byte rightIndex);
        public static Vector128<short> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector128<short> minuend, Vector128<short> left, Vector64<short> right, byte rightIndex);
        public static Vector128<int> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector128<int> minuend, Vector128<int> left, Vector64<int> right, byte rightIndex);
        public static Vector128<short> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector128<short> minuend, Vector128<short> left, Vector128<short> right, byte rightIndex);
        public static Vector128<int> RoundedDoublingMultiplySubtractBySelectedScalarSaturateHigh(Vector128<int> minuend, Vector128<int> left, Vector128<int> right, byte rightIndex);

        public static Vector64<short> RoundedDoublingMultiplySubtractSaturateHighScalar(Vector64<short> minuend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> RoundedDoublingMultiplySubtractSaturateHighScalar(Vector64<int> minuend, Vector64<int> left, Vector64<int> right);

        // TODO: Scalar BySelectedScalar variant

        /// <summary>
        /// Signed saturating Doubling Multiply-Add Long
        /// Corresponds to vector forms of SQDMULL and VQDMULL
        /// </summary>
        public static Vector64<short> DoublingMultiplyWideningLowerSaturate(Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplyWideningLowerSaturate(Vector64<int> left, Vector64<int> right);
        public static Vector128<short> DoublingMultiplyWideningLowerSaturate(Vector128<short> left, Vector128<short> right);
        public static Vector128<int> DoublingMultiplyWideningLowerSaturate(Vector128<int> left, Vector128<int> right);

        public static Vector64<short> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector64<short> left, Vector64<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector64<int> left, Vector64<int> right, byte rightIndex);
        public static Vector64<short> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector64<short> left, Vector128<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector64<int> left, Vector128<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector128<short> left, Vector64<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector128<int> left, Vector64<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector128<short> left, Vector128<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplyWideningLowerBySelectedScalarSaturate(Vector128<int> left, Vector128<int> right, byte rightIndex);

        public static Vector64<short> DoublingMultiplyWideningLowerSaturateScalar(Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplyWideningLowerSaturateScalar(Vector64<int> left, Vector64<int> right);

        // TODO: Scalar BySelectedScalar variant

        /// <summary>
        /// Signed saturating Doubling Multiply-Add Long
        /// Corresponds to vector forms of SQDMLAL and VQDMLAL
        /// </summary>
        public static Vector64<short> DoublingMultiplyAddWideningLowerSaturate(Vector64<short> addend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplyAddWideningLowerSaturate(Vector64<int> addend, Vector64<int> left, Vector64<int> right);
        public static Vector128<short> DoublingMultiplyAddWideningLowerSaturate(Vector128<short> addend, Vector128<short> left, Vector128<short> right);
        public static Vector128<int> DoublingMultiplyAddWideningLowerSaturate(Vector128<int> addend, Vector128<int> left, Vector128<int> right);

        public static Vector64<short> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector64<short> addend, Vector64<short> left, Vector64<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector64<int> addend, Vector64<int> left, Vector64<int> right, byte rightIndex);
        public static Vector64<short> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector64<short> addend, Vector64<short> left, Vector128<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector64<int> addend, Vector64<int> left, Vector128<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector128<short> addend, Vector128<short> left, Vector64<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector128<int> addend, Vector128<int> left, Vector64<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector128<short> addend, Vector128<short> left, Vector128<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplyAddWideningLowerBySelectedScalarSaturate(Vector128<int> addend, Vector128<int> left, Vector128<int> right, byte rightIndex);

        public static Vector64<short> DoublingMultiplyAddWideningLowerSaturateScalar(Vector64<short> addend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplyAddWideningLowerSaturateScalar(Vector64<int> addend, Vector64<int> left, Vector64<int> right);

        // TODO: Scalar BySelectedScalar variant

        /// <summary>
        /// Signed saturating Doubling Multiply-Subtract Long
        /// Corresponds to vector forms of SQDMLSL and VQDMLSL
        /// </summary>
        public static Vector64<short> DoublingMultiplySubtractWideningLowerSaturate(Vector64<short> minuend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplySubtractWideningLowerSaturate(Vector64<int> minuend, Vector64<int> left, Vector64<int> right);
        public static Vector128<short> DoublingMultiplySubtractWideningLowerSaturate(Vector128<short> minuend, Vector128<short> left, Vector128<short> right);
        public static Vector128<int> DoublingMultiplySubtractWideningLowerSaturate(Vector128<int> minuend, Vector128<int> left, Vector128<int> right);

        public static Vector64<short> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector64<short> minuend, Vector64<short> left, Vector64<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector64<int> minuend, Vector64<int> left, Vector64<int> right, byte rightIndex);
        public static Vector64<short> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector64<short> minuend, Vector64<short> left, Vector128<short> right, byte rightIndex);
        public static Vector64<int> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector64<int> minuend, Vector64<int> left, Vector128<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector128<short> minuend, Vector128<short> left, Vector64<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector128<int> minuend, Vector128<int> left, Vector64<int> right, byte rightIndex);
        public static Vector128<short> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector128<short> minuend, Vector128<short> left, Vector128<short> right, byte rightIndex);
        public static Vector128<int> DoublingMultiplySubtractWideningLowerBySelectedScalarSaturate(Vector128<int> minuend, Vector128<int> left, Vector128<int> right, byte rightIndex);

        public static Vector64<short> DoublingMultiplySubtractWideningLowerSaturateScalar(Vector64<short> minuend, Vector64<short> left, Vector64<short> right);
        public static Vector64<int> DoublingMultiplySubtractWideningLowerSaturateScalar(Vector64<int> minuend, Vector64<int> left, Vector64<int> right);

        // TODO: Scalar BySelectedScalar variant
    }
}
```

### LINKED ISSUE #37169: [Arm64] SIMD dot product
```c#
class Dp
{
  // int32x2_t vdot_s32 (int32x2_t r, int8x8_t a, int8x8_t b)
  //   A32: VSDOT.S8 Dd, Dn, Dm
  //   A64: SDOT Vd.2S, Vn.8B, Vm.8B
  public static Vector64<int> DotProduct(Vector64<int> addend, Vector64<sbyte> left, Vector64<sbyte> right);

  // uint32x2_t vdot_u32 (uint32x2_t r, uint8x8_t a, uint8x8_t b)
  //   A32: VUDOT.U8 Dd, Dn, Dm
  //   A64: UDOT Vd.2S, Vn.8B, Vm.8B
  public static Vector64<uint> DotProduct(Vector64<uint> addend, Vector64<byte> left, Vector64<byte> right);

  // int32x4_t vdotq_s32 (int32x4_t r, int8x16_t a, int8x16_t b)
  //   A32: VSDOT.S8 Qd, Qn, Qm
  //   A64: SDOT Vd.4S, Vn.16B, Vm.16B
  public static Vector128<int> DotProduct(Vector128<int> addend, Vector128<sbyte> left, Vector128<sbyte> right);

  // uint32x4_t vdotq_u32 (uint32x4_t r, uint8x16_t a, uint8x16_t b)
  //   A32: VUDOT.U8 Qd, Qn, Qm
  //   A64: UDOT Vd.4S, Vn.16B, Vm.16B
  public static Vector128<uint> DotProduct(Vector128<uint> addend, Vector128<byte> left, Vector128<byte> right);

  // int32x2_t vdot_lane_s32 (int32x2_t r, int8x8_t a, int8x8_t b, const int lane)
  //   A32: VSDOT.S8 Dd, Dn, Dm[lane]
  //   A64: SDOT Vd.2S, Vn.8B, Vm.4B[lane]
  public static Vector64<int> DotProductBySelectedQuadruplet(Vector64<int> addend, Vector64<sbyte> left, Vector64<sbyte> right, byte rightScaledIndex);

  // int32x2_t vdot_laneq_s32 (int32x2_t r, int8x8_t a, int8x16_t b, const int lane)
  //   A32: VSDOT.S8 Dd, Dn, Dm[lane]
  //   A64: SDOT Vd.2S, Vn.8B, Vm.4B[lane]
  public static Vector64<int> DotProductBySelectedQuadruplet(Vector64<int> addend, Vector64<sbyte> left, Vector128<sbyte> right, byte rightScaledIndex);

  // uint32x2_t vdot_lane_u32 (uint32x2_t r, uint8x8_t a, uint8x8_t b, const int lane)
  //   A32: VUDOT.U8 Dd, Dn, Dm[lane]
  //   A64: UDOT Vd.2S, Vn.8B, Vm.4B[lane]
  public static Vector64<uint> DotProductBySelectedQuadruplet(Vector64<uint> addend, Vector64<byte> left, Vector64<byte> right, byte rightScaledIndex);

  // uint32x2_t vdot_laneq_u32 (uint32x2_t r, uint8x8_t a, uint8x16_t b, const int lane)
  //   A32: VUDOT.U8 Dd, Dn, Dm[lane]
  //   A64: UDOT Vd.2S, Vn.8B, Vm.4B[lane]
  public static Vector64<uint> DotProductBySelectedQuadruplet(Vector64<uint> addend, Vector64<byte> left, Vector128<byte> right, byte rightScaledIndex);

  // int32x4_t vdotq_laneq_s32 (int32x4_t r, int8x16_t a, int8x16_t b, const int lane)
  //   A32: VSDOT.S8 Qd, Qn, Dm[lane]
  //   A64: SDOT Vd.4S, Vn.16B, Vm.4B[lane]
  public static Vector128<int> DotProductBySelectedQuadruplet(Vector128<int> addend, Vector128<sbyte> left, Vector128<sbyte> right, byte rightScaledIndex);

  // int32x4_t vdotq_lane_s32 (int32x4_t r, int8x16_t a, int8x8_t b, const int lane)
  //   A32: VSDOT.S8 Qd, Qn, Dm[lane]
  //   A64: SDOT Vd.4S, Vn.16B, Vm.4B[lane]
  public static Vector128<int> DotProductBySelectedQuadruplet(Vector128<int> addend, Vector128<sbyte> left, Vector64<sbyte> right, byte rightScaledIndex);

  // uint32x4_t vdotq_laneq_u32 (uint32x4_t r, uint8x16_t a, uint8x16_t b, const int lane)
  //   A32: VUDOT.U8 Qd, Qn, Dm[lane]
  //   A64: UDOT Vd.4S, Vn.16B, Vm.4B[lane]
  public static Vector128<uint> DotProductBySelectedQuadruplet(Vector128<uint> addend, Vector128<byte> left, Vector128<byte> right, byte rightScaledIndex);

  // uint32x4_t vdotq_lane_u32 (uint32x4_t r, uint8x16_t a, uint8x8_t b, const int lane)
  //   A32: VUDOT.U8 Qd, Qn, Dm[lane]
  //   A64: UDOT Vd.4S, Vn.16B, Vm.4B[lane]
  public static Vector128<uint> DotProductBySelectedQuadruplet(Vector128<uint> addend, Vector128<byte> left, Vector64<byte> right, byte rightScaledIndex);
}
```

This covers only `ARMv8.2-DotProd` but not the "Mixed sign integer dot product" (USDOT and SUDOT). My understanding that the latter should be under a different ISA class (e.g. MatrixMultiplication). Is my assumption correct?

@TamarChristinaArm @tannergooding @CarolEidt PTAL

The indexed form of these are weird. Here is a description from Arm Architecture Reference Manual
> This instruction performs the dot product of the four 8-bit elements in each 32-bit element of the first source register with the four 8-bit elements of an indexed 32-bit element in the second source register, accumulating the result into the corresponding 32-bit element of the destination register.

Another options how this could be expressed in the language is to use vector of ints for the second second source register. Then rightIndex does not have to be a scaled-index.
```c#
Vector128<uint> DotProductBySelectedQuadruplet(Vector128<uint> addend, Vector128<byte> left, Vector64<int> right, byte rightIndex)
```

But, then what should be used as a base type of the second operand for signed vs unsigned dot product? int vs uint?

I didn't like the second variant, so I decided to have index scaled, i.e. JIT will multiply the value by 4 and we expect the range of values for rightScaledIndex to be [0, 1] for `(Vector64<sbyte>/Vector64<byte> right)` and [0-3] for `(Vector128<sbyte>/Vector128<byte> right)`. 

Perhaps, someone can suggest better names for the methods or index operand?

### PR DESCRIPTION
**Pull request by echesakov**
* DotProduct

* DotProductBySelectedQuadruplet

* MultiplyRoundedDoublingAndAddSaturateHigh

* MultiplyRoundedDoublingAndAddSaturateHighScalar

* MultiplyRoundedDoublingAndSubtractSaturateHigh

* MultiplyRoundedDoublingAndSubtractSaturateHighScalar

* MultiplyRoundedDoublingBySelectedScalarAndAddSaturateHigh

* MultiplyRoundedDoublingBySelectedScalarAndSubtractSaturateHigh

* MultiplyRoundedDoublingScalarBySelectedScalarAndAddSaturateHigh

* MultiplyRoundedDoublingScalarBySelectedScalarAndSubtractSaturateHigh

Fixes #37169
Fixes #36696

With help from @eiriktsarpalis I collected the test results on an Arm64 device that supports these ISAs - I attached the archive with the files.

[Dp_Rdm_Rdm.Arm64.zip](https://github.com/dotnet/runtime/files/4898934/Dp_Rdm_Rdm.Arm64.zip)

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@CarolEidt @tannergooding @TamarChristinaArm This is ready for review, please take a look.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs
Comment (CarolEidt) [inline src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs:21]: I'm confused why this is here if there are no 64-bit intrinsics in this class.

File: src/coreclr/src/inc/corinfoinstructionset.h
Comment (tannergooding) [inline src/coreclr/src/inc/corinfoinstructionset.h:31]: Dp_Arm64 should also exist, it isn't covered by https://github.com/dotnet/runtime/pull/38460

**Review by CarolEidt — APPROVED** [review]
This LGTM overall, with one question. It seems that you and @tannergooding will have to coordinate with his #38460  (i.e. one will have to merge)

File: src/coreclr/src/inc/corinfoinstructionset.h
Comment (echesakovMSFT) [reply src/coreclr/src/inc/corinfoinstructionset.h:31]: Okay, I will update/rebase this PR after you merge #38460

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs
Comment (echesakovMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs:21]: I added this as a placeholder and I anticipated to wait till Tanner merges his PR and update thereafter

File: src/coreclr/src/pal/src/misc/jitsupport.cpp
Comment (tannergooding) [inline src/coreclr/src/pal/src/misc/jitsupport.cpp:53]: We should add a tracking issue with the `tracking-external-issue` label for when Windows support is in, so we can update the codeman checks for Windows.

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs
Comment (echesakovMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs:21]: My understanding that Dp.Arm64.IsSupported needs to return false if Dp.IsSupported is false.
While now it returns Dp.Arm64.IsSupported true if AdvSimd.Arm64.IsSupported is true, so we need to treat Dp.Arm64.IsSupported as intrinsics instead of being inherited from AdvSimd.Arm64.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs:21]: `Dp.Arm64.IsSupported` needs to return true if `Dp.IsSupported` and we are in a 64-bit process. It being `true` implies `Dp.IsSupported` and `ArmBase.Arm64.IsSupported` (which each also implies `ArmBase.IsSupported`).

Having it recursive here and adding the relevant entry to `Arm64VersionOfIsa` should hook it all up correctly (same as was done for Rdm: https://github.com/dotnet/runtime/pull/38957/files#diff-8e07abb3dff9bfe2e051a071c787f18bR27)

It exists because of https://github.com/dotnet/runtime/issues/34587 (and is the same reason as https://github.com/dotnet/runtime/pull/38460 was created), which is that we otherwise get into weird scenarios where `Dp.Arm64.IsSupported` can return true (because it actually resolves to `ArmBase.Arm64.IsSupported` when `Dp.IsSupported` returns false.

File: src/coreclr/src/inc/corinfoinstructionset.h
Comment (echesakovMSFT) [reply src/coreclr/src/inc/corinfoinstructionset.h]: Done

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs
Comment (echesakovMSFT) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs:21]: @CarolEidt Do you have any other questions here?

File: src/coreclr/src/pal/src/misc/jitsupport.cpp
Comment (echesakovMSFT) [reply src/coreclr/src/pal/src/misc/jitsupport.cpp:53]: Done https://github.com/dotnet/runtime/issues/39094

File: src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs
Comment (CarolEidt) [reply src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Arm/Dp.cs:21]: No, makes sense (sort of!)

File: src/coreclr/src/jit/hwintrinsicarm64.cpp
Comment (tannergooding) [inline src/coreclr/src/jit/hwintrinsicarm64.cpp:161]: Missing Dp_Arm64?

**Comment by tannergooding** [comment]
It would be good to update `JIT/Regression/JitBlue/Runtime_34587` to validate the hierarchy of the new ISAs

**Comment by echesakovMSFT** [comment]
> It would be good to update `JIT/Regression/JitBlue/Runtime_34587` to validate the hierarchy of the new ISAs

It looks that this tests was merged in the wrong place (under src/coreclr/tests/src instead of src/tests) - I will move it and update as you suggested

File: src/coreclr/src/jit/hwintrinsicarm64.cpp
Comment (echesakovMSFT) [reply src/coreclr/src/jit/hwintrinsicarm64.cpp:161]: Fixed

**Comment by echesakovMSFT** [comment]
@tannergooding Updated the test. Unfortunatelly, it does not do anything in CI now since our machines don't support Dp and Rdm

**Comment by tannergooding** [comment]
> Unfortunatelly, it does not do anything in CI now since our machines don't support Dp and Rdm

That's unfortunate, but possibly something we can address longer term. This scenario will be an issue (in the future) as we start onboarding newer ISAs

**Comment by echesakovMSFT** [comment]
@tannergooding I am going to merge this to unblock the remaining hardware intrinsics work.
Please let me know, if you want me to follow up on anything else here.

**Pull request merged**
